core/field.js

```diff
@@ -633,39 +638,114 @@ Blockly.Field.prototype.forceRerender = function() {
 };
 
 /**
- * By default there is no difference between the human-readable text and
- * the language-neutral values.  Subclasses (such as dropdown) may define this.
- * @return {string} Current value.
+ * Get the current value of the field.
+ * @return {*} Current value.
  */
 Blockly.Field.prototype.getValue = function() {
-  return this.getText();
+  return this.value_;
 };
 
 /**
- * By default there is no difference between the human-readable text and
- * the language-neutral values.  Subclasses (such as dropdown) may define this.
- * @param {string} newValue New value.
+ * Used to change the value of the field. Handles validation and events.
+ * Subclasses should override doClassValidation_ and doValueUpdate_ rather
+ * than this method.
+ * @param {*} newValue New value.
  */
 Blockly.Field.prototype.setValue = function(newValue) {
+  var doLogging = true;
   if (newValue === null) {
-    // No change if null.
+    doLogging && console.log('null, return');
+    // Not a valid value to check.
     return;
   }
-  // Validate input.
-  var validated = this.callValidator(newValue);
-  if (validated !== null) {
-    newValue = validated;
+
+  newValue = this.doClassValidation_(newValue);
+  if (newValue === null) {
+    doLogging && console.log('invalid, return');
+    this.doValueInvalid_();
+    if (this.isDirty_) {
+      this.forceRerender();
+    }
+    return;
+  }
+
+  var userValidator = this.getValidator();
+  if (userValidator) {
+    newValue = userValidator.call(this, newValue);
+    if (newValue === null) {
+      doLogging && console.log('invalid, return');
+      this.doValueInvalid_();
+      if (this.isDirty_) {
+        this.forceRerender();
+      }
+      return;
+    }
   }
-  // Check for change.
   var oldValue = this.getValue();
   if (oldValue == newValue) {
+    doLogging && console.log('same, return');
+    // No change.
     return;
   }
+
   if (this.sourceBlock_ && Blockly.Events.isEnabled()) {
     Blockly.Events.fire(new Blockly.Events.BlockChange(
         this.sourceBlock_, 'field', this.name, oldValue, newValue));
   }
-  this.setText(newValue);
+  this.doValueUpdate_(newValue);
+  if (this.isDirty_) {
+    // Call setText instead of calling forceRerender
+    // directly for backwards compatibility.
+    this.setText(newValue);
+    if (this.isDirty_) {
+      // setText must not have called forceRerender, either because it was
+      // overridden, or text_ was set inside doValueUpdate_. Try again.
+      this.forceRerender();
+    }
+  }
+  doLogging && console.log(this.value_);
+};
+
+/**
+ * A generic value possessed by the field.
+ * Should generally be non-null, only null when the field is created.
+ * @type {*}
+ * @protected
+ */
+Blockly.Field.prototype.value_ = null;
```

Comment:
Move this to the top of the file so it's not lost in here with the functions.

Reply:
Fixed